src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java

```diff
@@ -0,0 +1,950 @@
+        compactionCandidate = getCompactionCandidateBasedOnOverlaps(categorized);
+        if (compactionCandidate.isPresent())
+        {
+            return compactionCandidate.get();
+        }
+
+        return SSTablesForCompaction.EMPTY;
+    }
+
+    Optional<SSTablesForCompaction> getCompactionCandidateBasedOnSize(CategorizedSSTables categorizedSSTables)
+    {
+        double globalGarbage = getEstimatedGarbage(categorizedSSTables.stats);
+        double globalSizeOnDisk = categorizedSSTables.stats.onDiskLength;
+        logger.debug("{}: garbage: {}/{}", tableName, globalGarbage, globalSizeOnDisk);
+
+        // look for garbage cleanup.
+        List<OverlappingSet> overlappingTombstones = getDistinctOverlappingSSTables(categorizedSSTables.expiredTombstones);
+
+        Stream<DataCompactionCandidate> candidateStream = overlappingTombstones.stream()
+                .map(ot -> DataCompactionCandidate.get(categorizedSSTables.data, ot));
+
+        // if we are crossing the global threshold, just pick up the set with the most garbage, otherwise only choose the sets that are full of garbage.
+        // This is done so that many localized deletes can result in compaction.
+        candidateStream = isTooMuchGarbage(globalSizeOnDisk, globalGarbage, true)
+                ? candidateStream
+                : candidateStream.filter(c -> isTooMuchGarbage(c.stats.onDiskLength, c.estimatedGarbage, false));
+        Optional<DataCompactionCandidate> candidate = candidateStream.max(Comparator.comparingDouble(c -> c.estimatedGarbage));
+
+        if (!candidate.isPresent())
+        {
+            logger.debug("{}: no significant garbage found", tableName);
+            return Optional.empty();
+        }
+
+        List<SSTableStats> maxGarbageData = candidate.get().data;
+        OverlappingSet maxGarbageTombstones = candidate.get().tombstones;
+
+        TimeWindow dataSSTableTimeWindowMs = TimeWindow.merge(maxGarbageData.stream().map(s -> s.timeWindowMs).collect(Collectors.toList()));
+        if (logger.isDebugEnabled())
+        {
+            logger.debug("{}: candidate: {}, data_tw: {}", tableName, candidate.get(), dataSSTableTimeWindowMs);
+        }
+
+        // If there are too many files OR (data size is over the threshold and the data set is too wide),
+        // we need a sure shot way of reducing them.
+        if (maxGarbageTombstones.sstables.size() + maxGarbageData.size() > getMaxFileCountForCompaction() || dataSSTableTimeWindowMs.duration > 2 * getWindowSizeInMs())
+        {
+            if (maxGarbageTombstones.timeWindowMs.duration > 2 * getWindowSizeInMs() || maxGarbageTombstones.sstables.size() > 1)
+            {
+                logger.debug("{}: too many tombstone files", tableName);
+                return Optional.of(new SSTablesForCompaction(toSSTableReader(maxGarbageTombstones.sstables), true, true));
```

Comment:
Considering tombstoneMerge = true => the expired tombstones will not be present in resulting sstable(s),
and that these are ALL garbage tombstones, looks to me like we'll just delete a bunch of tombstones without deleting the corresponding data.